ext/TensorOperationscuTENSORExt.jl

```diff
@@ -1,350 +1,306 @@
+                            pAB::Index2Tuple,
+                            α::Number, β::Number, ::cuTENSORBackend)
+    # convert arguments
+    Ainds, Binds, Cinds = collect.(TO.contract_labels(pA, pB, pAB))
+    opA = tensorop(A, conjA)
+    opB = tensorop(B, conjB)
+
+    # dispatch to cuTENSOR
+    return contract!(α,
+                     A, Ainds, opA,
+                     B, Binds, opB,
+                     β,
+                     C, Cinds, OP_IDENTITY,
+                     OP_IDENTITY)
+end
 
-const cuTENSORBackend = TO.Backend{:cuTENSOR}
+function TO.tensortrace!(C::CuStridedView,
+                         A::CuStridedView, p::Index2Tuple, q::Index2Tuple, conjA::Symbol,
+                         α::Number, β::Number, ::cuTENSORBackend)
+    # convert arguments
+    Ainds, Cinds = collect.(TO.trace_labels(p, q))
+    opA = tensorop(A, conjA)
 
-function TO.tensoradd!(C::AbstractArray, pC::Index2Tuple,
-                       A::AbstractArray, conjA::Symbol, α::Number, β::Number,
-                       backend::cuTENSORBackend)
-    C_cuda = adapt(CuArray, C)
-    tensoradd!(C_cuda, pC, A, conjA, α, β, backend)
-    copyto!(C, collect(C_cuda))
-    return C
+    # map to reduction operation
+    plan = plan_trace(A, Ainds, opA, C, Cinds, OP_IDENTITY, OP_ADD)
+    return reduce!(plan, α, A, β, C)
 end
 
-function TO.tensoradd!(C::CuArray, pC::Index2Tuple,
-                       A::AbstractArray, conjA::Symbol, α::Number, β::Number,
-                       ::cuTENSORBackend)
-    return tensoradd!(C, pC, adapt(CuArray, A), conjA, α, β)
+function cuTENSOR.CuTensorDescriptor(a::CuStridedView; size=size(a), strides=strides(a),
+                                     eltype=eltype(a))
+    sz = collect(Int64, size)
+    st = collect(Int64, strides)
+    # compute largest possible alignment
+    alignment::UInt32 = 256
+    while (alignment > Base.aligned_sizeof(eltype)) && (alignment % 2 != 0)
+        alignment >>= 1
+    end
+    return cuTENSOR.CuTensorDescriptor(sz, st, eltype, alignment)
 end
 
-function TO.tensorcontract!(C::AbstractArray, pC::Index2Tuple,
-                            A::AbstractArray, pA::Index2Tuple, conjA::Symbol,
-                            B::AbstractArray, pB::Index2Tuple, conjB::Symbol,
-                            α::Number, β::Number, backend::cuTENSORBackend)
-    C_cuda = adapt(CuArray, C)
-    tensorcontract!(C_cuda, pC, A, pA, conjA, B, pB, conjB, α, β, backend)
-    copyto!(C, collect(C_cuda))
-    return C
-end
-function TO.tensorcontract!(C::CuArray, pC::Index2Tuple,
-                            A::AbstractArray, pA::Index2Tuple, conjA::Symbol,
-                            B::AbstractArray, pB::Index2Tuple, conjB::Symbol,
-                            α::Number, β::Number, ::cuTENSORBackend)
-    return tensorcontract!(C, pC, adapt(CuArray, A), pA, conjA, adapt(CuArray, B), pB,
-                           conjB, α, β)
+# trace!
+# ------
+# not actually part of cuTENSOR, just a special case of reduce
```

Comment:
Is this actually being used? In `TO.tensortrace!` you just call `reduce!` with the trace plan?

Reply:
Ah, I missed this. In principle this is jus a straight forward, so I'll just erase it